### PR TITLE
[Merged by Bors] - feat: Add lattice lemmas about `Sub{group,monoid}.{op,unop}`

### PR DESCRIPTION
**Pull request by eric-wieser**
In fact I only need the `closure` lemma, but the others are easy enough.

This changes the `opEquiv`s to be order isomorphisms rather than just `Equiv`s.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/Subgroup/MulOpposite.lean [hunk 1]
```diff
@@ -31,6 +31,13 @@ protected def op (H : Subgroup G) : Subgroup Gᵐᵒᵖ where
   mul_mem' ha hb := H.mul_mem hb ha
   inv_mem' := H.inv_mem
 
+@[to_additive (attr := simp)]
+theorem mem_op (x : Gᵐᵒᵖ) (S : Subgroup G) : x ∈ S.op ↔ x.unop ∈ S := Iff.rfl
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Subgroup/MulOpposite.lean:35]:
```suggestion
theorem mem_op {x : Gᵐᵒᵖ} {S : Subgroup G} : x ∈ S.op ↔ x.unop ∈ S := Iff.rfl
```

File: Mathlib/GroupTheory/Subgroup/MulOpposite.lean [hunk 2]
```diff
@@ -31,6 +31,13 @@ protected def op (H : Subgroup G) : Subgroup Gᵐᵒᵖ where
...
+
+@[to_additive (attr := simp, nolint simpNF)] lemma op_toSubmonoid (H : Subgroup G) :
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Subgroup/MulOpposite.lean:37]: What's the `simpNF` about?

File: Mathlib/GroupTheory/Subgroup/MulOpposite.lean [hunk 3]
```diff
@@ -40,25 +47,122 @@ protected def unop (H : Subgroup Gᵐᵒᵖ) : Subgroup G where
...
 /-- A subgroup `H` of `G` determines a subgroup `H.op` of the opposite group `Gᵐᵒᵖ`. -/
 @[to_additive (attr := simps) "An additive subgroup `H` of `G` determines an additive subgroup
 `H.op` of the opposite additive group `Gᵃᵒᵖ`."]
-def opEquiv : Subgroup G ≃ Subgroup Gᵐᵒᵖ where
+def opEquiv : Subgroup G ≃o Subgroup Gᵐᵒᵖ where
   toFun := Subgroup.op
   invFun := Subgroup.unop
   left_inv _ := SetLike.coe_injective rfl
   right_inv _ := SetLike.coe_injective rfl
+  map_rel_iff' := MulOpposite.op_surjective.forall
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Subgroup/MulOpposite.lean:89]:
```suggestion
  map_rel_iff' := op_le_op_iff
```

File: Mathlib/GroupTheory/Submonoid/MulOpposite.lean [hunk 4]
```diff
@@ -13,34 +13,138 @@ For every monoid `M`, we construct an equivalence between submonoids of `M` and
...
 
+@[to_additive (attr := simp)]
+theorem mem_op (x : Mᵐᵒᵖ) (S : Submonoid M) : x ∈ S.op ↔ x.unop ∈ S := Iff.rfl
+
 /-- Pull an opposite submonoid back to a submonoid along `MulOpposite.op`-/
 @[to_additive (attr := simps) "Pull an opposite additive submonoid back to a submonoid along
 `AddOpposite.op`"]
-protected def unop {M : Type*} [MulOneClass M] (x : Submonoid (MulOpposite M)) : Submonoid M where
-  carrier := MulOpposite.op ⁻¹' x.1
+protected def unop {M : Type*} [MulOneClass M] (x : Submonoid Mᵐᵒᵖ) : Submonoid M where
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Submonoid/MulOpposite.lean:34]:
```suggestion
protected def unop (x : Submonoid Mᵐᵒᵖ) : Submonoid M where
```

**Review by jcommelin — APPROVED** [review]
bors d+

File: Mathlib/GroupTheory/Subgroup/MulOpposite.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/GroupTheory/Subgroup/MulOpposite.lean]: Done throughout

File: Mathlib/GroupTheory/Subgroup/MulOpposite.lean [hunk 5]
```diff
@@ -40,25 +47,122 @@ protected def unop (H : Subgroup Gᵐᵒᵖ) : Subgroup G where
...
+theorem op_inf (S₁ S₂ : Subgroup G) : (S₁ ⊓ S₂).op = S₁.op ⊓ S₂.op := opEquiv.map_inf _ _
+
+@[to_additive]
+theorem unop_inf (S₁ S₂ : Subgroup Gᵐᵒᵖ) : (S₁ ⊓ S₂).unop = S₁.unop ⊓ S₂.unop :=
+  opEquiv.symm.map_inf _ _
+
+@[to_additive]
+theorem op_sSup (S : Set (Subgroup G)) : (sSup S).op = sSup (.unop ⁻¹' S) :=
+  (opEquiv.map_sSup _).trans <| sSup_image.symm.trans <|
+    congr_arg sSup <| opEquiv.image_eq_preimage _
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/GroupTheory/Subgroup/MulOpposite.lean]: Done, please check they look ok

bors d=@YaelDillies

**Comment by YaelDillies** [comment]
Thanks!

bors merge

**Pull request closed without merging**